Tab group tests

dhellgartner opened the pull request:
Added tests for sl-tab-group

Commit:
remove duplicate test

Commit:
Merge next branch

Merge remote-tracking branch 'origin/next' into tab-group-tests

Commit:
Add tests for sl-tab-group -- initial round of tests

Commit:
use individual fixtures for each test

Commit:
extract mocks + utility functions in external files

Commit:
Merge remote-tracking branch 'origin/next' into tab-group-tests

Commit:
remove unnecessary internals of intersection observer from the mock

Commit:
added first test on scroll buttons

Commit:
add scrolling tests

Commit:
remove resize observer mock

Resize observer is triggered but waiting for element
to be updated is not enough. You need to free the main thread
with the test for some time

Commit:
Also removed intersection observer mock

By waiting long enough for the things to happen automatically

Commit:
Merge next branch

Merge remote-tracking branch 'origin/next' into tab-group-tests

claviska commented on src/components/tab-group/tab-group.test.ts:
```diff
@@ -0,0 +1,291 @@
+  navigation?: DOMRect;
+}
+
+const wait = async (delayInMs = 0): Promise<void> => {
+  return new Promise(resolve => {
+    window.setTimeout(() => resolve(), delayInMs);
+  });
+};
```
There's a helper in `@open-wc/testing` called `aTimeout`. I believe it does the same as this.

claviska commented on src/components/tab-group/tab-group.test.ts:
```diff
@@ -0,0 +1,291 @@
+  return {};
+};
+
+const waitForScrollingToEnd = (element: Element, timeoutInMs = 500): Promise<void> => {
+  let lastLeft = element.scrollLeft;
+  let lastTop = element.scrollTop;
+  let framesWithoutChange = 0;
+  return new Promise((resolve, reject) => {
+    const timeout = window.setTimeout(() => {
+      reject(new Error('Waiting for scroll end timed out'));
+    }, timeoutInMs);
+    function checkScrollingChanged() {
+      if (element.scrollLeft !== lastLeft || element.scrollTop !== lastTop) {
+        framesWithoutChange = 0;
+        lastLeft = window.scrollX;
+        lastTop = window.scrollY;
+      } else {
+        framesWithoutChange++;
+        if (framesWithoutChange >= 20) {
+          clearTimeout(timeout);
+          resolve();
+        }
+      }
+      window.requestAnimationFrame(checkScrollingChanged);
+    }
+    checkScrollingChanged();
+  });
+};
```
This doesn't seem specific to tabs. Do you think it's worth moving into `src/internal/test.ts` or is it unlikely we'll use this anywhere else?

dhellgartner replied:
As `scrollIntoView` is used in also in `dropdown` and `select` it will probably be worth it --> extracted it

dhellgartner replied:
One note .... i am using an folder `src/internal/test/...` ... if you prefer having this al in `test.ts`, I do not have a strong opinion on that

claviska commented on src/components/tab-group/tab-group.test.ts:
```diff
@@ -0,0 +1,291 @@
+  });
+};
+
+const isElementVisibleFromScrolling = (outerElement: Element, innerElement: Element): boolean => {
+  const outerRect = outerElement.getBoundingClientRect();
+  const innerRect = innerElement.getBoundingClientRect();
+  return (
+    outerRect.top <= innerRect.bottom &&
+    innerRect.top <= outerRect.bottom &&
+    outerRect.left <= innerRect.right &&
+    innerRect.left <= outerRect.right
+  );
+};
```
Same here. Is this worth splitting out or not so much?

dhellgartner replied:
Extracted

claviska commented on src/internal/test/data-testid-helpers.ts:
```diff
@@ -0,0 +1,3 @@
+export const queryByTestId = <T extends Element>(container: HTMLElement, testId: string): T | null => {
+  return container.querySelector<T>(`[data-testid=${testId}]`);
```
Probably doesn't matter, but can't hurt just in case.

```suggestion
  return container.querySelector<T>(`[data-testid="${testId}"]`);
```

dhellgartner replied:
Done, thanks for that!

claviska commented on src/internal/test/data-testid-helpers.ts:
```diff
@@ -0,0 +1,3 @@
+export const queryByTestId = <T extends Element>(container: HTMLElement, testId: string): T | null => {
```
Just curious about this because you mentioned having interest in working on more tests later on — is this something we're going to be using in many tests? If so, a JSDoc comment to help future devs understand what it does would be great! 😄

dhellgartner replied:
At least I tend to use that a lot and have come to like it ;) . So I added a JSDoc comment (also for the other extracted helper functions)

claviska reviewed (changes requested):
This looks really good, nice and clean! 😄 

I left a few comments but overall this is just about perfect. Thank you for your effort on this!

Let me know if you have any questions!

Commit:
Merge changes from next branch

Merge remote-tracking branch 'origin/next' into tab-group-tests

Commit:
Fix problems with resize observer

These problems appeared after npm ci but (according
to the sources linked in the comments) unproblematic

Commit:
Handle merge request comments

* replace custom wait function with corresponding function
 from openwc/testing
* Extracted waitForScrollingToEnd and isElementVisibleFromScrolling into
dedicated files to be reused
* Improve queryByTestId --> make it usable for more complex values
* Add js docs

dhellgartner commented:
@claviska How would you prefere to go on with this merge request:
* The branch has been open for quite some time
* On the other hand there is still some functionality missing to be tested

So do you prefer to keep this open till that all has been solved or merge back as soon as all checks are green and open a new branc/MR afterwards?

Commit:
run lint fix

claviska commented:
> So do you prefer to keep this open till that all has been solved or merge back as soon as all checks are green and open a new branc/MR afterwards?

We can keep it open until you're ready. Just make sure to unmark it as draft and ping me when you're looking to move forward. I don't want it to get lost in the noise. 😄

Commit:
Added tests for selecting a tab by click

Commit:
added further tests for tab group selection

Commit:
Merge remote-tracking branch 'origin/next' into tab-group-tests

claviska commented on src/components/tab-group/tab-group.test.ts:
```diff
@@ -0,0 +1,450 @@
+  name: string;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const waitForScrollButtonsToBeRendered = (tabGroup: SlTabGroup): Promise<any> => {
+  return waitUntil(() => {
+    const scrollButtons = tabGroup.shadowRoot?.querySelectorAll('sl-icon-button');
+    return scrollButtons?.length === 2;
+  });
+};
```
It doesn't look like we use the promise's return value. Should we drop the return and use `Promise<void>` to avoid the eslint rule?

dhellgartner replied:
No, that is a mistake. I fixed it.

> Thanks again for all your work on this.

No worries, I learned a lot when doing this. Thank you for the helpful comments.

claviska reviewed (approved):
Thanks again for all your work on this. The tests are clean and easy to understand. I'm happy to merge this when you're ready.

I did leave one comment, but it's up to you if you want to update anything or leave it as-is.

Commit:
use Promise<void> instead of Promise<any>

to avoid eslint errors

claviska commented:
I just noticed that the tab group test is dumping two `null` lines out. 

<img width="403" alt="CleanShot 2023-02-07 at 15 55 04@2x" src="https://user-images.githubusercontent.com/55639/217363583-9078e25f-d582-4599-913d-aeabb85d64a4.png">

I've narrowed it down to the two tests that call `waitForScrollButtonsToBeRendered()`. Any idea why this is happening and how we can fix it?

dhellgartner commented:
Will have a look as soon as I have time

dhellgartner commented:
I found the issue causing the null logs:
* web test library installs a global event handler in `node_modules/@web/browser-logs/dist/logUncaughtErrors.js`
* Additionally, the Resize observer sometimes throws errors which are nothing to worry about, see also the corresponding comment on tab-group.test.ts. Note that this error event is somewhat strange as it does not have an associated error, so `errorEvent.error === null`. The event handler mentioned above trying to log the error causes the `null` log lines.
* Unfortunately, the error handler installed from web test library takes precedence wrt. the error handlers installed in the tests as it was installed first
* To fix this, I needed to install an error handler even earlier which is quite an invasive change.
* I create a PR for that (https://github.com/shoelace-style/shoelace/pull/1196) but I do have full understanding if you want to close it as the change is pretty harsh ... I just did not find any other method to avoid the null messages.
